Un-unicode output for Windows CMD

MinchinWeb opened the pull request:
Fixes #119

CleanCut commented on green/test/test_windows.py:
```diff
@@ -23,3 +23,6 @@ def test_colorOutput(self):
                         colorama.ansitowin32.StreamWrapper))
 
+    def test_unicode_override(self):
+        pass
```
Looking pretty good, except for this poor test here. ;-)

MinchinWeb replied:
It is a poor, sad test :(

I couldn't figure out how to get `green` to run a `unittest` testcase, and then validate what gets printed out. The [actual testcase](https://github.com/CleanCut/green/issues/119#issuecomment-209138184) is nearly trivial...

```
class UnicodeTest(unittest.TestCase):
    def test1(self):
        print('\u03BB')  # lambda
        self.assertTrue(True)
```

Suggestions on how to go about this?

jayvdb replied:
The true test is to spawn subprocess of green running that test method, and check the output.

https://github.com/wikimedia/pywikibot-core/blob/master/tests/pwb_tests.py#L71

There all sorts of gotchas, especially when doing this on Windows, and only Python versions, resulting a mess like
https://github.com/wikimedia/pywikibot-core/blob/master/tests/utils.py#L771

However I suspect you'll only need to set `PYTHONIOENCODING` to utf-8 for the subprocess

jayvdb commented:
All of these "on_appveyor" conditions look like handling only the simple case of Appveyor, and dont really test using green on Windows because there is special logic for Appveyor only.

e.g. I expect that `green > green.log` on Windows will fail, because it is more like an Appveyor environment (i.e. no stdin, which Python uses to determine console encoding).

This is just a general comment, as I realise that this changeset isnt introducing "on_appveyor", but it is making it more "normal" by exposing it as an attribute.

jayvdb commented on green/output.py:
```diff
@@ -144,14 +145,16 @@ class GreenStream(object):
         # colors, but it WILL interpret posix ansi escape codes!
-        on_windows = platform.system() == 'Windows'
-        on_appveyor = os.environ.get('APPVEYOR', False)
+        self.on_windows = platform.system() == 'Windows'
```
this shouldnt be an instance variable.  it should be a package level constant.

jayvdb commented on green/output.py:
```diff
@@ -144,14 +145,16 @@ class GreenStream(object):
-        on_windows = platform.system() == 'Windows'
-        on_appveyor = os.environ.get('APPVEYOR', False)
+        self.on_windows = platform.system() == 'Windows'
+        self.on_appveyor = os.environ.get('APPVEYOR', False)
```
likewise, I expect this should be a package level constant.  I dont see how this env var could ever be expected to change during an execution of `green`.

Also, wherever it is placed, it should be a private attribute, as this is not something that should become part of the public `green` contract.

MinchinWeb commented:
@jayvdb I've reworked the 'flags' inside the module a little to try and make things more explicit.

The test failed on Travis because `coverage` timed out.

The [test failed](https://ci.appveyor.com/project/CleanCut/green/build/303) on Appvoyor on a test where colorama is supposed to strip ANSI codes. But the issue is something gets fed a `str` but is expecting a `unicode`. Is this an issue with Python 2.7? With `colorama`? With `unidecode`? I'm not sure. Part of the reason I moved to Python 3 was to avoid the str/unicode mess...  Suggestions??

In any case, the tests do pass on my local machine.

CleanCut commented on green/output.py:
```diff
@@ -167,6 +171,9 @@ def writeln(self, text=''):
             text = text.decode('utf-8')
+        # Compensate for windows' anti-social unicode behavior
+        if self._ascii_only_output:
+            text = unidecode(text)
```
Sorry for the delay.  I have been just absolutely swamped.

I suggest just changing line 176 to read:

    text = unidecode(unicode(text))

That may be work and we could be done with this and merge it.

CleanCut replied:
Hmm, wait.  Python 3 doesn't have the unicode() function anymore.

Since it's a Python 2-specific fix, I suppose we should make that condition:

    if self._ascii_only_output:
        if platform.python_version_tuple()[0] == '2': # pragma: no cover
           text = unicode(text)
        text = unidecode(text)

MinchinWeb commented:
@CleanCut: I've made the changes you suggested

CleanCut commented:
Included in the 2.4.2 release.  Which was just released.
